api/api.py

```diff
@@ -5,59 +5,66 @@
 
 
-def verify_coordinates_were_provided(
-    origin_lat, origin_lon, destination_lat, destination_lon
-):
+def verify_coordinates_were_provided(origin, destination):
     """Ensure all origin/destination coordinates were provided."""
-    return origin_lat and origin_lon and destination_lat and destination_lon
+    return (
+        origin
+        and destination
+        and origin["latitude"]
+        and origin["longitude"]
+        and destination["latitude"]
+        and destination["longitude"]
+    )
 
 
 def get_co2_estimate(
     transport_mode=None,
+    origin=None,
+    destination=None,
     distance_km=None,
     vehicle_occupancy=None,
-    origin_lat=None,
-    origin_lon=None,
-    destination_lat=None,
-    destination_lon=None,
+    *args,
+    **kwargs,
 ):
-    """Entry point for connexion as specified by operationId in specification.json"""
+    """Estimate CO2 for a given leg, preserving original properties"""
 
-    coordinates_were_provided = verify_coordinates_were_provided(
-        origin_lat, origin_lon, destination_lat, destination_lon
-    )
+    coordinates_were_provided = verify_coordinates_were_provided(origin, destination)
 
     if distance_km is not None:
         # default to using provided distance
         pass
     elif coordinates_were_provided:
         # Calculate distance from origin/destination
-        origin = (origin_lat, origin_lon)
-        destination = (destination_lat, destination_lon)
+        origin = (origin["latitude"], origin["longitude"])
+        destination = (destination["latitude"], destination["longitude"])
+
         distance_km = haversine(origin, destination)
     else:
-        return (
-            {"error": "Not enough information was provided to calculate CO2 estimate."},
-            400,
-        )
+        return {
+            "error_message": "Not enough information was provided to calculate CO2 estimate."
+        }
 
-    if isinstance(transport_mode, str):
-        transport_mode = Mode[transport_mode]
+    mode = Mode[transport_mode]
 
     if vehicle_occupancy is None:
-        vehicle_occupancy = transport_mode.average_occupancy
+        vehicle_occupancy = mode.avg_occupancy
 
-    co2_estimate = estimate_co2(
-        mode=transport_mode, distance_in_km=distance_km, occupancy=vehicle_occupancy
+    co2_estimate = mode.estimate_co2(
+        distance_in_km=distance_km, occupancy=vehicle_occupancy
     )
 
-    return_data = {
-        "transport_mode": transport_mode.name
-        if isinstance(transport_mode, Mode)
-        else transport_mode,
+    return {
+        "transport_mode": transport_mode,
         "vehicle_occupancy": vehicle_occupancy,
         "distance_km": distance_km,
         "co2_estimate": co2_estimate,
     }
 
-    return (return_data, 200)
+
+def estimate_co2(body):
+    # Join the CO2 estimate object with original request object
+    co2_estimates = [
+        {**get_co2_estimate(**leg), **{"request_object": leg}} for leg in body
```

Comment:
That second unpacking is overkill, just `{**get_co2_estimate(**leg), "request_object": leg}` will do.

Reply:
It's worth mentioning that if any `leg` happens to have a `request_object` key itself, it'd be overwritten here, so it might be worth making it more "hidden", e.g., `_request_object`.

Reply:
I have now updated the code to make sure there is no collision between properties on the `co2_estimate` and `request_object`:

```py
# Create a dictionary with CO2 estimate and original request object
co2_estimates = [
    {"co2_estimate": get_co2_estimate(**leg), "request_object": leg} for leg in body
]
```